velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,345 @@
+    return credentials_;
+  }
+};
+
+class GCSFileSystem::Impl {
+ public:
+  Impl(const Config* config) : gcsConfig_(config) {}
+
+  ~Impl() {}
+
+  // Use the input Config parameters and initialize the GCSClient.
+  void initializeClient() {
+    auto options = gc::Options{};
+    std::string scheme = gcsConfig_.scheme();
+    // if (scheme.empty()) scheme = "https";
+    if (scheme == "https") {
+      options.set<gc::UnifiedCredentialsOption>(
+          gc::MakeGoogleDefaultCredentials());
+    } else {
+      options.set<gc::UnifiedCredentialsOption>(gc::MakeInsecureCredentials());
+    }
+    options.set<gcs::UploadBufferSizeOption>(kUploadBufferSize);
+    if (!gcsConfig_.endpointOverride().empty()) {
+      options.set<gcs::RestEndpointOption>(
+          scheme + "://" + gcsConfig_.endpointOverride());
+    }
+    if (gcsConfig_.credentials()) {
+      options.set<gc::UnifiedCredentialsOption>(gcsConfig_.credentials());
+    }
+    client_ = std::make_shared<gcs::Client>(options);
+  }
+
+  gcs::Client* gcsClient() const {
+    return client_.get();
+  }
+
+ private:
+  const GCSConfig gcsConfig_;
+  std::shared_ptr<gcs::Client> client_;
+};
+
+GCSFileSystem::GCSFileSystem(std::shared_ptr<const Config> config)
+    : FileSystem(config) {
+  impl_ = std::make_shared<Impl>(config.get());
+}
+
+void GCSFileSystem::initializeClient() {
+  impl_->initializeClient();
+}
+
+std::unique_ptr<ReadFile> GCSFileSystem::openFileForRead(
+    std::string_view path) {
+  const std::string gcspath = gcsPath(path);
+  auto gcsfile = std::make_unique<GCSReadFile>(gcspath, impl_->gcsClient());
+  gcsfile->initialize();
+  return gcsfile;
+}
+
+std::unique_ptr<WriteFile> GCSFileSystem::openFileForWrite(
+    std::string_view path) {
+  const std::string gcspath = gcsPath(path);
+  auto gcsfile = std::make_unique<GCSWriteFile>(gcspath, impl_->gcsClient());
+  gcsfile->initialize();
+  return gcsfile;
+}
+
+void GCSFileSystem::remove(std::string_view path) {
+  if (!isGCSFile(path))
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+  const std::string file = gcsPath(path);
+
+  // assumption it's a proper path
+  std::string bucket;
+  std::string object;
+  bucketAndKeyFromGCSPath(file, bucket, object);
+
+  if (!object.empty()) {
+    auto stat = impl_->gcsClient()->GetObjectMetadata(bucket, object);
+    VELOX_CHECK_GCS_OUTCOME(
+        stat.status(), "Failed to get metadata for GCS object", bucket, object);
+  }
+  auto ret = impl_->gcsClient()->DeleteObject(bucket, object);
+  VELOX_CHECK_GCS_OUTCOME(
+      ret, "Failed to get metadata for GCS object", bucket, object);
+}
+
+std::string GCSFileSystem::name() const {
+  return "GCS";
+}
+
+folly::once_flag GCSInstantiationFlag;
+static std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
+    filesystemGenerator = [](std::shared_ptr<const Config> properties) {
+      // Only one instance of GCSFileSystem is supported for now (follow S3 for
```

Comment:
S3 API is supported by at least 2 storage systems (Minio, AWS S3). So we might need to support two instances in the future.
However, GCS is only supported by Google, correct? Will we ever need two instances?

Reply:
I guess we will need multiple instances for different credentials.